nginx.tmpl

```diff
@@ -500,79 +457,161 @@ proxy_set_header X-Original-URI $request_uri;
 proxy_set_header Proxy "";
 {{- end }}
 
-{{- /* Precompute some information about each vhost. */}}
+{{- /* Precompute and store some information about vhost that use VIRTUAL_HOST_MULTIPORTS. */}}
+{{- range $vhosts_yaml, $containers := groupBy $globals.containers "Env.VIRTUAL_HOST_MULTIPORTS" }}
+    {{- range $hostname, $vhost := (fromYaml $vhosts_yaml) }}
+        {{- $vhost_data := when (hasKey $globals.vhosts $hostname) (get $globals.vhosts $hostname) (dict) }}
+        {{- $paths := coalesce $vhost_data.paths (dict) }}
+        
+        {{- if (empty $vhost) }}
+            {{ $vhost = dict "/" (dict) }}
+        {{- end }}
+
+        {{- range $path, $vpath := $vhost }}
+            {{- if (empty $vpath) }}
+                {{- $vpath = dict "dest" "" "port" "default" }}
+            {{- end }}
+            {{- $dest := coalesce $vpath.dest "" }}
+            {{- $port := when (hasKey $vpath "port") (toString $vpath.port) "default" }}
+            {{- $path_data := when (hasKey $paths $path) (get $paths $path) (dict) }}
+            {{- $path_ports := when (hasKey $path_data "ports") (get $path_data "ports") (dict) }}
+            {{- $path_port_containers := when (hasKey $path_ports $port) (get $path_ports $port) (list) }}
+            {{- $path_port_containers = concat $path_port_containers $containers }}
+            {{- $_ := set $path_ports $port $path_port_containers }}
+            {{- $_ := set $path_data "ports" $path_ports }}
+            {{- if (not (hasKey $path_data "dest")) }}
+                {{- $_ := set $path_data "dest" $dest }}
+            {{- end }}
+            {{- $_ := set $paths $path $path_data }}
```

Comment:
@pini-gh this is the only part I still have some difficulty wrapping my brain around, do you think you could describe what the code is doing here ?

If I understood correctly we're updating `$globals.vhosts` in both loop in a way that something like that:

```yaml
services:
  web1-a:
    image: foo
    environment:
      VIRTUAL_HOST: web1.example.com
      VIRTUAL_PORT: 9000

  web1-b:
    image: foo
    environment:
      VIRTUAL_HOST_MULTIPORT: |-
        web1.example.com:
          "/":
            port: 9000
```

actually put the two containers in the same `upstream` ?

Reply:
> ```yaml
> services:
>   web1-a:
>     image: foo
>     environment:
>       VIRTUAL_HOST: web1.example.com
>       VIRTUAL_PORT: 9000
> 
>   web1-b:
>     image: foo
>     environment:
>       VIRTUAL_HOST_MULTIPORT: |-
>         web1.example.com:
>           "/":
>             port: 9000
>     
> ```
> actually put the two containers in the same `upstream` ?

Indeed. What identify precisely the upstream is the pair [ hostname, path ]. If makes no sense to have several upstreams with the same pair. Only one can be used in the end, because 'hostname' maps to a unique `server` block, and 'path' to a unique `location` in that block.

Reply:
👍 neat. I guess we should add a test for that, then.

Do you see other cases that should be tested before we can consider merging this feature ?

Reply:
The main concern is that this syntax alone allows defining conflicting setups, and the template should be resilient to that.

Reply:
Well, one actually : same 'path' with different 'dest'. The others I thought about are avoided by the data structure using the 'hostname' and the 'path' as keys.

Remain the cases where:
* the YAML synctax is not correct => ignore it and print a big fat warning into the generated file ?
* or where the same key 'hostname' or 'path' is found several times in the same dict => I don't know if it makes the parsing fail or if the last declaration silently override the previous ones.

Reply:
> the YAML synctax is not correct => ignore it and print a big fat warning into the generated file ?

Seems easy, working on it.

The other option would be to use `mustFromYaml` which panick when not able to parse the provided string, but that won't be super resilient as docker-gen currently don't really handle this and then panick itself.

> or where the same key 'hostname' or 'path' is found several times in the same dict => I don't know if it makes the parsing fail or if the last declaration silently overrides the previous ones.

Tested: parsing fail, so it will be caught by the above check and warning.

> Well, one actually : same 'path' with different 'dest'.

Do you mean something like this ?

```yaml
services:
  web1-a:
    image: foo
    environment:
      VIRTUAL_HOST_MULTIPORT: |-
        web1.example.com:
          "/":
            port: 9000

  web1-b:
    image: foo
    environment:
      VIRTUAL_HOST_MULTIPORT: |-
        web1.example.com:
          "/":
            port: 9000
            dest: "/bar"
```

Reply:
> Do you mean something like this ?

Yes.

Reply:
I just tested it, what happen is that the second containers `dest` is silently discarded here:
```
{{- if (not (hasKey $path_data "dest")) }}
    {{- $_ := set $path_data "dest" $dest }}
{{- end }}
```
because `$globals.vhosts` already has a `dest` set for this hostname / path from the first container.

Reply:
That's the same behavior that with VIRTUAL_HOST / VIRTUAL_PATH / VIRTUAL_DEST when different VIRTUAL_DEST are defined for the same VIRTUAL_HOST / VIRTUAL_PATH pair.

So I think this is OK.